Add Gradient Fill Support to Canvas Backend

Atharva-Kanherkar opened the pull request:
###  **Changes Made:**  
- Implemented `GradientFill` for Canvas backend.  
- Supported Linear and Radial gradients using `ctx.createLinearGradient()` and `ctx.createRadialGradient()`.  
 

## **Solves** #167

Atharva-Kanherkar commented:
I'll fix the failing tests.

noelwelsh commented:
You have changed the formatting of the file and deleted the comments. Please don't do this. It makes it very difficult to see what you have changed as the diffs show every line in the file has changed, and losing the comments makes it harder for future developers. 

If you run the `build` command from sbt, as described at https://www.creativescala.org/doodle/development/process.html, that will take care of all the formatting as well as adding headers and so on.

Atharva-Kanherkar commented:
> You have changed the formatting of the file and deleted the comments. Please don't do this. It makes it very difficult to see what you have changed as the diffs show every line in the file has changed, and losing the comments makes it harder for future developers.
> 
> If you run the `build` command from sbt, as described at https://www.creativescala.org/doodle/development/process.html, that will take care of all the formatting as well as adding headers and so on.

I do not know how did this happen. I did run the build command from sbt and then only committed the code when it fully compiled. Let me try it again, and sorry for the inconvenience!

Atharva-Kanherkar commented:
Hi, @noelwelsh . I tried formatting it manually, even after running the build command and also added back the comments that were missing. If it still fails, do you want me to redo everything again or is there another way to solve the formatting issue?

noelwelsh commented:
There are still some formatting issues but it's much clearer now what you have added. I can fix the remaining formatting issues later.

Can you delete the file `circle.pdf` that you have committed? I assume this was created by a test, or by you playing around with Doodle, and should not be part of this PR. 

Are you planning to add any tests, such as converting the gradient examples to use the Canvas backend?

Atharva-Kanherkar commented:
> There are still some formatting issues but it's much clearer now what you have added. I can fix the remaining formatting issues later.
> 
> Can you delete the file `circle.pdf` that you have committed? I assume this was created by a test, or by you playing around with Doodle, and should not be part of this PR.
> 
> Are you planning to add any tests, such as converting the gradient examples to use the Canvas backend?

Okay sure! I will delete the file, i think it got added because of the tests. 
Yes, I would love to do that. Do you want me to open another PR for that? And do you want me to add them in the examples folder?

noelwelsh commented:
>  Yes, I would love to do that. Do you want me to open another PR for that? And do you want me to add them in the examples folder?

Please add them to this PR. If will the PR a complete unit of work, and it will also be easier to review as the tests will validate the correctness of the code.

Atharva-Kanherkar commented:
> > Yes, I would love to do that. Do you want me to open another PR for that? And do you want me to add them in the examples folder?
> 
> Please add them to this PR. If will the PR a complete unit of work, and it will also be easier to review as the tests will validate the correctness of the code.

Okay! Could you guide me on writing the tests? Even a little information to get started would be helpful!

noelwelsh commented:
It turns out that the examples for the Canvas backend already include the gradients. You can see them yourself if you look at the documentation in `/docs/target/docs/site/canvas/style-examples.html`, which will be built by the `build` command or `docs / tlSite` if you don't want to go through all the other steps the `build` command does.

If you look at the examples you'll see the linear gradient is correct but the radial gradient has an error somewhere as it does not display.

The examples include the gradients are defined here: https://github.com/creativescala/doodle/blob/main/examples/js/src/main/scala/doodle/examples/StyleExamples.scala#L26

They are rendered for the Canvas here: https://github.com/creativescala/doodle/blob/main/examples/js/src/main/scala/doodle/examples/canvas/CanvasStyleExamples.scala

They are included in the Canvas documentation here: https://github.com/creativescala/doodle/blob/main/docs/src/pages/canvas/style-examples.md

Atharva-Kanherkar commented:
> It turns out that the examples for the Canvas backend already include the gradients. You can see them yourself if you look at the documentation in `/docs/target/docs/site/canvas/style-examples.html`, which will be built by the `build` command or `docs / tlSite` if you don't want to go through all the other steps the `build` command does.
> 
> If you look at the examples you'll see the linear gradient is correct but the radial gradient has an error somewhere as it does not display.
> 
> The examples including the gradients are defined here: https://github.com/creativescala/doodle/blob/main/examples/js/src/main/scala/doodle/examples/StyleExamples.scala#L26
> 
> They are rendered for the Canvas here: https://github.com/creativescala/doodle/blob/main/examples/js/src/main/scala/doodle/examples/canvas/CanvasStyleExamples.scala
> 
> They are included in the Canvas documentation here: https://github.com/creativescala/doodle/blob/main/docs/src/pages/canvas/style-examples.md

Okay sure! I will look into it!

noelwelsh commented:
> I have been trying to run the `docs/tlSite` command, but, even after having insalled npm and node and it being in the valid path, I am getting these errors 

Looks like a problem with your installation of npm. This is not something I can really help you with as I don't know how your system is setup. It doesn't particularly matter though. npm is only used to build the CSS. If you look at [the definition of the `tlSite` command](https://github.com/creativescala/doodle/blob/main/build.sbt#L178) you'll see that it consists of a number of commands. You can run all the commands but the `css` command you should get usable output.

>  I was trying to navigate to `/docs/target/docs/site/canvas/style-examples.html` but this was giving some issues, how can I fix these?

You haven't described the issues so I have no way of helping you.

Atharva-Kanherkar commented:
> > I have been trying to run the `docs/tlSite` command, but, even after having insalled npm and node and it being in the valid path, I am getting these errors
> 
> Looks like a problem with your installation of npm. This is not something I can really help you with as I don't know how your system is setup. It doesn't particularly matter though. npm is only used to build the CSS. If you look at [the definition of the `tlSite` command](https://github.com/creativescala/doodle/blob/main/build.sbt#L178) you'll see that it consists of a number of commands. You can run all the commands but the `css` command you should get usable output. So, from within `sbt`, something like the following:
> 
> ```
>  examples.js / Compile / fastLinkJS
>  docs / Compile / run
>  docs / mdoc
>  docs / laikaSite
> ```
> 
> > I was trying to navigate to `/docs/target/docs/site/canvas/style-examples.html` but this was giving some issues, how can I fix these?
> 
> You haven't described the issues so I have no way of helping you.

 I'm facing issues with `docs / laikaSite`, particularly with `npx tailwindcss`, despite npm working fine. I tried running commands individually and adjusting CSS paths, but no luck.  

Is there an alternative way to test my code locally, or should I push the changes and rely on CI?  

 ![image](https://github.com/user-attachments/assets/e3beb0b3-f9d7-4ea3-a912-3dd6f6229419)

Thanks!

noelwelsh commented:
Ok, so the issue is the CSS file is missing, and Laika complains about that. You have two options I can see:

* Generate the CSS yourself, by running the same command that the `css` task does, but in a shell where npm works. You can copy the command straight out of `build.sbt`. You'll probably need to run `npm install` first.
* Copy the CSS from [the Doodle website](https://www.creativescala.org/doodle/creative-scala.css) and put it into the correct place so that the rest of the process finds that file and proceeds normally.  I think if you look at the `css` command in `build.sbt` that will show you the directory to place the file into.

I don't think you can use CI as I don't know of any way to view the HTML output from the CI runs without publishing the site, which we don't want to do until everything is working correctly.

Atharva-Kanherkar commented:
> Ok, so the issue is the CSS file is missing, and Laika complains about that. You have two options I can see:
> 
> * Generate the CSS yourself, by running the same command that the `css` task does, but in a shell where npm works. You can copy the command straight out of `build.sbt`. You'll probably need to run `npm install` first.
> * Copy the CSS from [the Doodle website](https://www.creativescala.org/doodle/creative-scala.css) and put it into the correct place so that the rest of the process finds that file and proceeds normally.  I think if you look at the `css` command in `build.sbt` that will show you the directory to place the file into.
> 
> I don't think you can use CI as I don't know of any way to view the HTML output from the CI runs without publishing the site, which we don't want to do until everything is working correctly.

I have been trying all these methods. I also tried moving the file to another directory which the error was pointing to. But no luck. I have tried a lot, But I still do not find the root issue where is it :( 
![image](https://github.com/user-attachments/assets/6064904b-6294-4da0-9413-4cb61fe4c045)

Atharva-Kanherkar commented:
> It's very strange that Laika would not find the CSS file if you placed it in the correct directory. Anyway, if you can't get it working you can just delete the line that imports the stylesheet, which is the one causing the errors, from the template. The HTML output will be unstyled but it should still be usable for the purpose of inspecting the Canvas output.

![image](https://github.com/user-attachments/assets/96c6dce9-2d5e-43c1-b949-363017a3f270)
 Thank you! It happened. Now I can proceed

noelwelsh commented:
Great! That was much more work than I thought it would, but I'm glad it's now fixed and you can proceed.

Atharva-Kanherkar commented:
> Great! That was much more work than I thought it would be, but I'm glad it's now fixed and you can proceed with fixing the problem.

Yes it was kind of difficult with laika, but glad it worked out. Had exams past week so couldn't work a lot on this :) Now I will try to finish this up today!

Atharva-Kanherkar commented:
Hello @noelwelsh! It seems that the radial gradient does display correctly here, as shown in the image. Given this, I guess no changes to the file are necessary. What do you think would be the best next step?


![image](https://github.com/user-attachments/assets/d183b6cd-8de0-403d-b602-6be4c1fccab7)

noelwelsh commented:
> Hello @noelwelsh! It seems that the radial gradient does display correctly here, as shown in the image. Given this, I guess no changes to the file are necessary. What do you think would be the best next step?

Interesting. That's not the output I saw. Perhaps there is an issue on different browsers. What browser are you using? I was using Firefox. Perhaps you can check the output on different browsers.

Atharva-Kanherkar commented:
> > Hello @noelwelsh! It seems that the radial gradient does display correctly here, as shown in the image. Given this, I guess no changes to the file are necessary. What do you think would be the best next step?
> 
> Interesting. That's not the output I saw. Perhaps there is an issue on different browsers. What browser are you using? I was using Firefox. Perhaps you can check the output on a few different browsers?
> 
> Update: I checked out your repository and tested on both Firefox and Safari. Neither showed the radial gradient. I suspect the coordinates are incorrect. I suspect you'd see the same for the linear gradient if the linear gradient ran vertically.

Was using brave ( a fork of chrome ). I will try seeing this on firefox, and I will update you if I see the gradients!

Atharva-Kanherkar commented:
Hi, @noelwelsh .  These are firefox and brave's results side by side. 
![image](https://github.com/user-attachments/assets/1e74cc6b-70d8-46d9-b971-3dcf472bd242)
As you can see, they look exactly the same. Could you let me know the exact coordinates you suspect might be incorrect? I'd be happy to investigate further.

Atharva-Kanherkar commented:
Hey @noelwelsh! Just wanted to check in on the next steps. Let me know if there's anything you'd like me to update or clarify. Thanks for your time!

noelwelsh commented:
I haven't been able to get the Canvas output working on my computer, though I'm going to make another attempt at this. Can you create a few more examples of gradients to see if they show any errors?

There are also a few issues with the PR: 

* CI is failing
* you have included a file, circle.pdf, that doesn't belong as part of the PR.

noelwelsh commented:
The issue is an incorrect inner radius is passed to `createRadialGradient`. Tested on Chrome, Firefox, and Safari.

I don't know why I am seeing this and you are not. Perhaps you have changes you have not pushed?

Atharva-Kanherkar commented:
> The issue is an incorrect inner radius is passed to `createRadialGradient`. Tested on Chrome, Firefox, and Safari.
> 
> I don't know why I am seeing this and you are not. Perhaps you have changes you have not pushed?

I checked and there WERE some changes I did not push. However, I was wondering that once you give me a green flag if the main functionality of the issue that we were solving works, I will remove the circle.pdf and also fix the ci. I think I did not push some changes, but now that they are pushed, I think it should work fine now. If you give me a green flag, Ill fix the CI!

noelwelsh commented:
I am disappointed with the lack of attention to detail in your work. 

In your latest commit you have again added changes that should not be part of the PR. I think this is the third time you have done this. You need to pay attention to what you add to a commit, and not just blindly commit every change in every file.

Additionally, you have made an arbitrary choice for the value of the inner radius. Why is it `radius / 4`, and not, say, `radius / 8`? If you created additional test cases to compare the output to, say, SVG output, you would see that this is not the correct value. If you read, say, the SVG code handling gradients, you'd find what the correct value is.

This change only requires about 10 lines of code. The challenge is not in writing the code, but in doing the work to figure out what those changes should be. So far you not showing much capability for independent work.

Atharva-Kanherkar commented:
> I am disappointed with the lack of attention to detail in your work.
> 
> In your latest commit you have again added changes that should not be part of the PR. I think this is the third time you have done this. You need to pay attention to what you add to a commit, and not just blindly commit every change in every file.
> 
> Additionally, you have made an arbitrary choice for the value of the inner radius. Why is it `radius / 4`, and not, say, `radius / 8`? If you created additional test cases to compare the output to, say, SVG output, you would see that this is not the correct value. If you read, say, the SVG code handling gradients, you'd find what the correct value is.
> 
> This change only requires about 10 lines of code. The challenge is not in writing the code, but in doing the work to figure out what those changes should be. So far you not showing much capability for independent work.

 I am very sorry you felt like this. It wasn't my intention to disappoint you. The issues must have happened due to my lack of careful attention to commits.

Firstly, I apologize for what has happened. I should have been more diligent in reviewing my commits before pushing.

Secondly, I have removed the circle.pdf file in the next commit.

Regarding the radius/4 usage: I've changed it to 0 because that's the correct implementation for radial gradients. The inner radius should be 0 to properly match how radial gradients work in other backends like SVG, where gradients start from a point source (effectively radius 0) and expand outward. Using radius/4 was incorrect as it created a gradient starting from a circle with 1/4 the final radius rather than from a point, which doesn't match the expected behavior of standard radial gradients.

The HTML Canvas API's createRadialGradient(x0, y0, r0, x1, y1, r1) should use r0=0 to create the point-to-circle effect that matches the mathematical definition of a radial gradient and aligns with the implementation in other rendering backends.

Please Correct me if I am wrong. I was exploring the project, and it must have been that I might have missed something. I will try my best not to repeat that if i solve some other issues.

noelwelsh commented:
The task is to implement the missing functionality in the `setFill` method in `CanvasDrawing.scala`. Your PR includes unrelated changes to `CanvasDrawing.scala` and unrelated changes to three other files. I cannot accept this PR while it contains these unrelated changes. Either remove these changes or, if that is too difficult, create a new PR containing just the changes to `setFill`.

You are correct that the inner radius should be zero. My expectation is that you would work this out yourself using the strategies I suggested: by comparing a range of examples to the gradient implementations on other backends or by reading the documentation for the methods used on other backends.

Atharva-Kanherkar commented:
> The task is to implement the missing functionality in the `setFill` method in `CanvasDrawing.scala`. Your PR includes unrelated changes to `CanvasDrawing.scala` and unrelated changes to three other files. I cannot accept this PR while it contains these unrelated changes. Either remove these changes or, if that is too difficult, create a new PR containing just the changes to `setFill`.
> 
> You are correct that the inner radius should be zero. My expectation is that you would work this out yourself using the strategies I suggested: by comparing a range of examples to the gradient implementations on other backends or by reading the documentation for the methods used on other backends.

For sure. I will do that. I am sorry for the problems caused. Will rectify them ASAP.

Atharva-Kanherkar commented:
> The task is to implement the missing functionality in the `setFill` method in `CanvasDrawing.scala`. Your PR includes unrelated changes to `CanvasDrawing.scala` and unrelated changes to three other files. I cannot accept this PR while it contains these unrelated changes. Either remove these changes or, if that is too difficult, create a new PR containing just the changes to `setFill`.
> 
> You are correct that the inner radius should be zero. My expectation is that you would work this out yourself using the strategies I suggested: by comparing a range of examples to the gradient implementations on other backends or by reading the documentation for the methods used on other backends.

I am making a new pull request and closing this one for convenience.
